src/extent.c

```diff
@@ -306,80 +313,70 @@ extents_npages_get(extents_t *extents) {
-	 * don't need an atomic fetch-add; we can get by with a load followed by
-	 * a store.
-	 */
-	size_t cur_extents_npages =
-	    atomic_load_zu(&extents->npages, ATOMIC_RELAXED);
-	atomic_store_zu(&extents->npages, cur_extents_npages + npages,
+	atomic_fetch_add_zu(&extents->npages, npages,
 	    ATOMIC_RELAXED);
+	mutex_pool_unlock(tsdn, &arena->extents_mutex_pool, pind);
 }
 
 static void
-extents_remove_locked(tsdn_t *tsdn, extents_t *extents, extent_t *extent,
-    bool preserve_lru) {
-	malloc_mutex_assert_owner(tsdn, &extents->mtx);
+extents_remove_locked(tsdn_t *tsdn, arena_t *arena, extents_t *extents, extent_t *extent) {
 	assert(extent_state_get(extent) == extents->state);
 
 	size_t size = extent_size_get(extent);
 	size_t psz = extent_size_quantize_floor(size);
 	pszind_t pind = sz_psz2ind(psz);
 	extent_heap_remove(&extents->heaps[pind], extent);
 	if (extent_heap_empty(&extents->heaps[pind])) {
+		malloc_mutex_lock(tsdn, &extents->bitmap_mtx);
 		bitmap_set(extents->bitmap, &extents_bitmap_info,
 		    (size_t)pind);
+		malloc_mutex_unlock(tsdn, &extents->bitmap_mtx);
 	}
-	if (!preserve_lru) {
-		extent_list_remove(&extents->lru, extent);
-	}
+	extent_list_remove(&extents->lru[pind], extent);
+
 	size_t npages = size >> LG_PAGE;
-	/*
-	 * As in extents_insert_locked, we hold extents->mtx and so don't need
-	 * atomic operations for updating extents->npages.
-	 */
-	size_t cur_extents_npages =
-	    atomic_load_zu(&extents->npages, ATOMIC_RELAXED);
-	assert(cur_extents_npages >= npages);
-	atomic_store_zu(&extents->npages,
-	    cur_extents_npages - (size >> LG_PAGE), ATOMIC_RELAXED);
+	assert(atomic_load_zu(&extents->npages, ATOMIC_RELAXED) >= npages);
+	atomic_fetch_sub_zu(&extents->npages, npages, ATOMIC_RELAXED);
 }
 
 /*
  * Find an extent with size [min_size, max_size) to satisfy the alignment
  * requirement.  For each size, try only the first extent in the heap.
  */
 static extent_t *
-extents_fit_alignment(extents_t *extents, size_t min_size, size_t max_size,
-    size_t alignment) {
-        pszind_t pind = sz_psz2ind(extent_size_quantize_ceil(min_size));
-        pszind_t pind_max = sz_psz2ind(extent_size_quantize_ceil(max_size));
+extents_fit_alignment(tsdn_t *tsdn, arena_t * arena, extents_t *extents,
+		      size_t min_size, size_t max_size,
+		      size_t alignment) {
+	pszind_t pind = sz_psz2ind(extent_size_quantize_ceil(min_size));
+	pszind_t pind_max = sz_psz2ind(extent_size_quantize_ceil(max_size));
 
 	for (pszind_t i = (pszind_t)bitmap_ffu(extents->bitmap,
 	    &extents_bitmap_info, (size_t)pind); i < pind_max; i =
 	    (pszind_t)bitmap_ffu(extents->bitmap, &extents_bitmap_info,
 	    (size_t)i+1)) {
 		assert(i < NPSIZES);
-		assert(!extent_heap_empty(&extents->heaps[i]));
+		mutex_pool_lock(tsdn, &arena->extents_mutex_pool, i);
 		extent_t *extent = extent_heap_first(&extents->heaps[i]);
+		if (!extent) {
+			mutex_pool_unlock(tsdn, &arena->extents_mutex_pool, i);
```

Comment:
nit: change the two unlock+continue to a goto the unlock at the end of the loop?

also, change to `if (extent != NULL) `, and a few other places with `if (!extent)` or `if (extent)`.